src/cascadia/TerminalSettingsModel/FileUtils.cpp

```diff
@@ -183,9 +157,70 @@ namespace winrt::Microsoft::Terminal::Settings::Model
+            // single entry (ACE) -- a mandatory label (i.e. a
+            // LABEL_SECURITY_INFORMATION) that sets the file integrity level to
+            // "high",  with a no-write-up policy.
+            //
+            // When accessed from a security context at a lower integrity level,
+            // the no-write-up policy filters out rights that aren't in the
+            // object type's generic read and execute set (for the file type,
+            // that's FILE_GENERIC_READ | FILE_GENERIC_EXECUTE).
+            //
+            // Another option we considered here was manually setting the ACLs
+            // on this file such that Builtin\Admins could read&write the file,
+            // and all users could only read.
+            //
+            // Big thanks to @eryksun in GH#11222 for helping with this. This
+            // alternative method was chosen because it's considerably simpler.
+
+            // The required security descriptor can be created easily from the
+            // SDDL string: "S:(ML;;NW;;;HI)"
+            // (i.e. SACL:mandatory label;;no write up;;;high integrity level)
+            unsigned long cb;
+            THROW_IF_WIN32_BOOL_FALSE(
+                ConvertStringSecurityDescriptorToSecurityDescriptor(L"S:(ML;;NW;;;HI)",
+                                                                    SDDL_REVISION_1,
+                                                                    wil::out_param_ptr<PSECURITY_DESCRIPTOR*>(sd),
+                                                                    &cb));
+
+            // Initialize a security attributes structure.
+            sa.nLength = sizeof(SECURITY_ATTRIBUTES);
+            sa.lpSecurityDescriptor = sd.get();
+            sa.bInheritHandle = false;
+
+            // If we're running in an elevated context, when this file is
+            // created, it will automatically be owned by
+            // Builtin\Administrators, which will pass the above
+            // _isOwnedByAdministrators check.
+            //
+            // Programs running in an elevated context will be free to write the
+            // file, and unelevated processes will be able to read the file. An
+            // unelevated process could always delete the file and rename a new
+            // file in it's place (a la the way `vim.exe` saves files), but if
+            // they do that, the new file _won't_ be owned by Administrators,
+            // failing the above check.
+        }
+
+        wil::unique_hfile file{ CreateFileW(path.c_str(),
+                                            GENERIC_WRITE,
+                                            FILE_SHARE_READ | FILE_SHARE_DELETE,
+                                            elevatedOnly ? &sa : nullptr,
+                                            CREATE_ALWAYS,
+                                            FILE_ATTRIBUTE_NORMAL,
+                                            nullptr) };

```

Comment:
The existing security is retained when the file is overwritten with `CREATE_ALWAYS` disposition. That's okay only if the file is owned by the administrators group. To check this, maybe change `_isOwnedByAdministrators()` to take a file handle and call `GetSecurityInfo()` instead of `GetNamedSecurityInfoW()`. This entails also updating `ReadUTF8File()` to move the check after `CreateFileW()`. (Related: I think `ReadUTF8File` should be able to do the file create outside of the loop, and just reset the file pointer on each pass.)

In `WriteUTF8File()`, if `elevatedOnly` is true and the file isn't owned by the administrators group, or opening fails with `ERROR_ACCESS_DENIED`, I suggest doing an atomic replace. Create a new file beside the existing file, with a unique name and `CREATE_NEW` disposition. Then rename the new file to the target name. 

---

An alternative to an atomic replace would be to rewrite the security of the existing file, but that's more complicated. The existing permissions may not allow opening with `WRITE_OWNER` and `WRITE_DAC` access, or even `GENERIC_WRITE` access. SeRestorePrivilege would have to be enabled for the current thread, and the open would have to use `FILE_FLAG_BACKUP_SEMANTICS`. The security info to set would be `OWNER_SECURITY_INFORMATION | LABEL_SECURITY_INFORMATION | DACL_SECURITY_INFORMATION | UNPROTECTED_DACL_SECURITY_INFORMATION`. Setting "unprotected" DACL information allows using an empty DACL (not `NULL`; just empty), since the permissions will be inherited from the parent directory. This avoids having to create a custom DACL for the current user.

Reply:
This problem with `WriteUTF8File()` should be avoided, else the write is just a waste of time since `ReadUTF8File()` will ignore it. Personally, I'd wrap `CreateFileW()` with the addition of the `elevatedOnly` parameter. Replace an existing file with a new empty file if the existing file isn't owned by the administrators. An administrator should be free to change anything about the file security except for the owner. There's no way to know whether to trust the file if it's not owned by the administrators group.

Reply:
I've read your comments just now, but I'm not sure I understand them entirely...
First of all: Is there a security issue with the current code?

The way I understand this, we don't check the security info during writing which can cause us to create a file with the wrong security info despite `elevatedOnly` being true, right? But is that an issue if we delete that file during `ReadUTF8File` anyways? I personally don't mind this being a "waste of time", because this would be a situation that can only be created artificially, for instance by a user creating the file themselves... As such it's unlikely to occur in practice, isn't it?

Reply:
>  Is there a security issue with the current code?

I don't think so because the user always has read access anyway. There's just not any point in writing data that will be thrown away.

> this would be a situation that can only be created artificially, for instance by a user 
> creating the file themselves

Yes, for example it would occur if the user edits the file using an editor that deletes the original before saving. (Overwriting a file just deletes its data, alternate data streams, and extended attributes, but not its security descriptor.) This will succeed in a standard security context, since the file is in a directory that grants the user `FILE_DELETE_CHILD` access. The new file will be owned by the user instead of the administrators group and inherit its permissions from the parent directory.